### PR TITLE
add proxy settings for kube up with vagrant as provider

### PR DESCRIPTION
**Pull request by mfanjie**
The PR contains two changes:
1). enable proxy settings for kube-up.sh when KUBERNETES_PROVIDER=vagrant, also updates getings start docs
This would require users to install vagrant proxy config plugin and export environment variables for proxy settings.

2). update proxy settings environment variables name for Kubernetes build to align with others with format e.g. KUBERNETES_HTTP_PROXY(originally was KUBE_BUILD_HTTP_PROXY). so that users only need to set the proxy for build and cluster up with same environment variables.

more discussions in #21529

### CODE REVIEW

File: build/build-image/Dockerfile [hunk 1]
```diff
@@ -15,9 +15,9 @@
 # This file creates a standard build environment for building Kubernetes
 FROM gcr.io/google_containers/kube-cross:KUBE_BUILD_IMAGE_CROSS_TAG
 
-ENV http_proxy=KUBE_BUILD_HTTP_PROXY \
-  https_proxy=KUBE_BUILD_HTTPS_PROXY \
-  no_proxy=KUBE_BUILD_NO_PROXY
+ENV http_proxy=KUBERNETES_HTTP_PROXY \
```
Comment (ixdy) [inline build/build-image/Dockerfile]: do we still need these? I don't think we're doing any network access inside the build image anymore.

**Comment by ixdy** [comment]
ok to test

**Comment by mfanjie** [comment]
@ixdy you are right, the golang package download was move to cross images, so it is not required anymore in build/build-image/Dockerfile.
BTW, if found minor issue in sed_opts part, if we define the options as array (-i \'\')or string("-i ''"), a backup file Dockerfile'' will be generated on Mac, so I changed it to plain format.

File: build/common.sh [hunk 2]
```diff
@@ -242,14 +242,10 @@ function kube::build::is_osx() {
 
 function kube::build::update_dockerfile() {
   if kube::build::is_osx; then
-    sed_opts=("-i ''")
```
Comment (ixdy) [inline build/common.sh]: I think I figured out how to make this work: remove the double quotes here, but add them when you expand. i.e.

```shell
if kube::build::is_osx:
  sed_opts=(-i '')
else
  sed_opts=(-i)
fi
sed "${sed_opts[@]}" "s/KUBE_BUILD_IMAGE_CROSS_TAG/${KUBE_BUILD_IMAGE_CROSS_TAG}/" "${LOCAL_OUTPUT_BUILD_CONTEXT}/Dockerfile"
```

**Comment by ixdy** [comment]
other than that nit, basically LGTM, I think.

@derekwaynecarr do you have any comments on the Vagrant stuff?

File: docs/getting-started-guides/vagrant.md [hunk 3]
```diff
@@ -412,6 +412,24 @@ export KUBERNETES_MASTER_MEMORY=1536
 export KUBERNETES_NODE_MEMORY=2048
 ```
 
+#### I want to set proxy settings for my Kubernetes cluster boot strapping!
+
+If you are behind of a proxy, you need to install vagrant proxy plugin and set the proxy settings by
```
Comment (derekwaynecarr) [inline docs/getting-started-guides/vagrant.md]: If you are behind a proxy...

File: docs/getting-started-guides/vagrant.md [hunk 4]
```diff
@@ -412,6 +412,24 @@ export KUBERNETES_MASTER_MEMORY=1536
...
+
+If you are behind of a proxy, you need to install vagrant proxy plugin and set the proxy settings by 
+
+```sh
+vagrant plugin install vagrant-proxyconf
+export KUBERNETES_HTTP_PROXY=http://username:password@proxyaddr:proxyport
+export KUBERNETES_HTTP_PROXY=https://username:password@proxyaddr:proxyport
+```
+
+Optionally, you can specify addresses of no proxy for kubernetes build, for example
```
Comment (derekwaynecarr) [inline docs/getting-started-guides/vagrant.md]: This reads awkwardly to me... 

Optionally you can specify addresses to not proxy?

**Comment by derekwaynecarr** [comment]
A few nits on the documentation, other than that LGTM

File: build/common.sh (hunk 2 above)
Comment (mfanjie) [inline build/common.sh]: yep, this is working correctly, I made the change, thanks.

**Comment by mfanjie** [comment]
@derekwaynecarr your comments addressed, thanks.

File: docs/getting-started-guides/vagrant.md [hunk 5]
```diff
@@ -412,6 +412,24 @@ export KUBERNETES_MASTER_MEMORY=1536
...
 ```
 
+#### I want to set proxy settings for my Kubernetes cluster boot strapping!
+
+If you are behind a proxy, you need to install vagrant proxy plugin and set the proxy settings by 
+
+```sh
+vagrant plugin install vagrant-proxyconf
+export KUBERNETES_HTTP_PROXY=http://username:password@proxyaddr:proxyport
+export KUBERNETES_HTTP_PROXY=https://username:password@proxyaddr:proxyport
```
Comment (ixdy) [inline docs/getting-started-guides/vagrant.md]: this should probably be `KUBERNETES_HTTPS_PROXY`? (note S)

**Comment by ixdy** [comment]
Sorry for delay in review - basically LG besides the minor documentation nit (not sure how I missed that earlier).

Please also squash your commits now. Thanks!

@k8s-bot unit test this again please, issue #IGNORE

**Comment by mfanjie** [comment]
Sorry, I was out yesterday, I squashed the commit and fixed the typo.

**Comment by ixdy** [comment]
Looks like docs need to be fixed:
```
Verifying ./hack/../hack/verify-generated-docs.sh
+++ [0229 18:01:40] Building go targets for linux/amd64:
    cmd/gendocs
    cmd/genkubedocs
    cmd/genman
    cmd/genbashcomp
    cmd/mungedocs
+++ [0229 18:02:17] Placing binaries
/workspace/kubernetes/docs/getting-started-guides/vagrant.md
----
remove-whitespace:
not printing failed chunk: too many lines
contents were modified

FAIL: changes needed but not made due to --verify
/workspace/kubernetes/docs/ is out of date. Please run hack/update-generated-docs.sh
!!! Error in ./hack/../hack/verify-generated-docs.sh:28
  '"${KUBE_ROOT}/hack/after-build/verify-generated-docs.sh" "$@"' exited with status 1
Call stack:
  1: ./hack/../hack/verify-generated-docs.sh:28 main(...)
Exiting with status 1
```

also, did you force-push the squash? I see 4 commits now.

**Comment by mfanjie** [comment]
I will clean things up then resubmit. Sorry for the mess.

**Comment by ixdy** [comment]
No worries. If you can't reuse this PR anymore, feel free to open a new one and cc @ me for review.

A tip: in general I find using `git merge --ff-only` or `git rebase` to pull in upstream changes to my dev branches to work best for keeping things clean, and then use `git rebase -i` to squash commits for final merge.

**Pull request closed without merging**